mesa/experimental/cell_space/property_layer.py

```diff
@@ -0,0 +1,392 @@
+
+        If a NumPy ufunc is used, an additional value should be provided.
+
+        Args:
+            operation: A function to apply. Can be a lambda function or a NumPy ufunc.
+            value: The value to be used if the operation is a NumPy ufunc. Ignored for lambda functions.
+            condition: (Optional) A callable that returns a boolean array when applied to the data.
+        """
+        condition_array = np.ones_like(
+            self._mesa_data, dtype=bool
+        )  # Default condition (all cells)
+        if condition is not None:
+            vectorized_condition = np.vectorize(condition)
+            condition_array = vectorized_condition(self._mesa_data)
+
+        # Check if the operation is a lambda function or a NumPy ufunc
+        if isinstance(operation, np.ufunc):
+            if ufunc_requires_additional_input(operation):
+                if value is None:
+                    raise ValueError("This ufunc requires an additional input value.")
+                modified_data = operation(self._mesa_data, value)
+            else:
+                modified_data = operation(self._mesa_data)
+        else:
+            # Vectorize non-ufunc operations
+            vectorized_operation = np.vectorize(operation)
+            modified_data = vectorized_operation(self._mesa_data)
+
+        self._mesa_data = np.where(condition_array, modified_data, self._mesa_data)
+
+    def select_cells(self, condition: Callable, return_list=True):
+        """Find cells that meet a specified condition using NumPy's boolean indexing, in-place.
+
+        # fixme: consider splitting into two separate functions
+        #  select_cells_boolean
+        #  select_cells_index
+
+        Args:
+            condition: A callable that returns a boolean array when applied to the data.
+            return_list: (Optional) If True, return a list of (x, y) tuples. Otherwise, return a boolean array.
+
+        Returns:
+            A list of (x, y) tuples or a boolean array.
+        """
+        condition_array = condition(self._mesa_data)
+        if return_list:
+            return list(zip(*np.where(condition_array)))
+        else:
+            return condition_array
+
+    def aggregate(self, operation: Callable):
+        """Perform an aggregate operation (e.g., sum, mean) on a property across all cells.
+
+        Args:
+            operation: A function to apply. Can be a lambda function or a NumPy ufunc.
+        """
+        return operation(self._mesa_data)
+
+
+class HasPropertyLayers:
+    """Mixin-like class to add property layer functionality to Grids."""
+
+    # fixme is there a way to indicate that a mixin only works with specific classes?
+    def __init__(self, *args, **kwargs):
+        """Initialize a HasPropertyLayers instance."""
+        super().__init__(*args, **kwargs)
+        self._mesa_property_layers = {}
+
+    def create_property_layer(
```

Comment:
Can we let `create_property_layer` return the PropertyLayer also? That would allow users to directly assign it and manipulate, if they want to do so.

It doesn't have to be a keyword, because `self.grid.create_property_layer()` will also still work if the return is not used.